server/docs/design/bidi-producer-consumers-streaming.md

```diff
@@ -0,0 +1,339 @@
+transmit a continuous flow of BlockItem messages while the HTTP/2 connection is open.
+
+**Producer StreamObserver** - The Producer StreamObserver is a custom implementation of the [gRPC StreamObserver 
+interface](https://github.com/grpc/grpc-java/blob/0ff3f8e4ac4c265e91b4a0379a32cf25a0a2b2f7/stub/src/main/java/io/grpc/stub/StreamObserver.java#L45) used by Helidon.  It is initialized by the BlockItemStreamService (see Entities section).  Helidon invokes the Producer
+StreamObserver at runtime when the producer sends a new BlockItem to the `StreamSink` gRPC service. 
+
+**Consumer StreamObserver** - The Consumer StreamObserver is a custom implementation of the [gRPC StreamObserver 
+interface](https://github.com/grpc/grpc-java/blob/0ff3f8e4ac4c265e91b4a0379a32cf25a0a2b2f7/stub/src/main/java/io/grpc/stub/StreamObserver.java#L45) used by Helidon.  It is initialized by the BlockItemStreamService (see Entities section).  Helidon invokes the Consumer
+StreamObserver at runtime when the downstream consumer of the `StreamSource` gRPC service sends HTTP/2 responses to 
+sent BlockItems. 
+
+**subscribe** - Consumers calling the `StreamSource` gRPC service must be affiliated or subscribed with a producer to 
+receive a live stream of BlockItems from the `hedera-block-node`.
+
+**unsubscribe** - Consumers terminating their connection with the `StreamSource` gRPC service must be unaffiliated or 
+unsubscribed from a producer so that internal objects can be cleaned up and resources released.
+
+---
+
+### Block Node gRPC Streaming Services API
+
+The following protobuf definition outlines the gRPC services and messages used to stream BlockItems between a producer,
+consumers and the `hedera-block-node`.  The `BlockStreamGrpc` service definition provides 2 bidirectional streaming
+methods: `StreamSink` and `StreamSource`.  Aside from the gRPC service and method names, all the types will be replaced 
+by those outlined in [hedera-protobufs](https://github.com/hashgraph/hedera-protobufs/pull/342/files).
+
+```protobuf
+
+/**
+  * The BlockStreamGrpc service definition provides 2 bidirectional streaming methods for
+  * exchanging BlockItems with the Block Node server.
+  *
+  * A producer (e.g. Consensus Node) can use the StreamSink method to stream BlockItems to the
+  * Block Node server.  The Block Node server will respond with a BlockResponse message for
+  * each BlockItem received.
+  *
+  * A consumer (e.g. Mirror Node) can use the StreamSource method to request a stream of
+  * BlockItems from the server.  The consumer is expected to respond with a BlockResponse message
+  * with the id of each BlockItem received.
+ */
+service BlockStreamGrpc {
+  /**
+  * StreamSink is a bidirectional streaming method that allows a producer to stream BlockItems 
+  * to the Block Node server.  The server will respond with a BlockResponse message for each
+  * BlockItem received.
+  */
+  rpc StreamSink (stream BlockItem) returns (stream BlockItemResponse) {}
```

Comment:
I don't love the Source and Sink names in API. Saying that open to better ideas `StreamIn` and `StreamOut` are first that come to mind. @rbair23 thoughts?

Reply:
Should the names reflect additional properties of the methods to leave room for future definitions?  For example, `UnverifiedStreamOut` to leave room for another method called `VerifiedStreamOut`?

Reply:
I'm not a fan of putting `in` or `out` in rpc verbs.
We currently have the RPC verbs defined as follows for publishing a stream to the block node and subscribing to a stream from the block node:
```protobuf
    rpc publishBlockStream (stream PublishStreamRequest) returns (stream PublishStreamResponse);
    rpc subscribeBlockStream(stream SubscribeStreamRequest) returns (stream SubscribeStreamResponse);
```

Reply:
I removed the API definition and replaced it with a link to the hedera-protobufs repo and the branch where to find the correct service definitions and types.